File: tests/vasca_runners/vasca_run.m
function vasca_run(varargin)
    % This script runs the vasca function and saves the results to a JSON file.
    %
    % It takes command-line arguments to specify the input data and optional parameters:
    %
    % Usage: vasca_run <parglmVS> <siglev>
    %   - <parglmVS>: The output structure from the parglmVS function (in JSON format).
    %   - <siglev>: Significance level (0.01 by default).
    %
    % Example usage in MATLAB:
    %   matlab -nodisplay -r "vasca_run('parglmVS_output.json', 0.05)"
    %   matlab -nodisplay -r "vasca_run('parglmVS_output.json')"  % uses default siglev of 0.01
    %
    % Outputs:
    %   - Saves the results as 'vasca_results.json'
    
    % Get command-line arguments
    if isoctave()
        args = argv();
    else
        args = varargin;
    end
    
    % Check number of arguments (at least parglmVS must be provided)
    if numel(args) < 1
        error(['Not enough arguments. Usage: vasca_run <parglmVS> <siglev>\n']);
    end
    
    % Parse parglmVS (path to the JSON file)
    argParglmVS = args{1};
    if exist(argParglmVS, 'file') == 2
        % It's a file, try to load it as a JSON file
        try
            fid = fopen(argParglmVS, 'r');
            if fid == -1
                error('Unable to open file: %s', argParglmVS);
            end
            raw = fread(fid, inf);
            str = char(raw');
            fclose(fid);
            parglmoVS = jsondecode(str); % Decode the JSON content
            fprintf('Loaded parglmVS from: %s\n', argParglmVS);
        catch ME
            error('Error loading parglmVS from JSON file: %s', ME.message);
        end
    else
        error('parglmVS must be a valid file.');
    end

    % Parse siglev (if provided, otherwise use default 0.01)
    if numel(args) > 1
        siglev = str2double(args{2});
        if isnan(siglev)
            error('The significance level must be a valid number.');
        end
        fprintf('Using significance level: %f\n', siglev);
    else
        siglev = 0.01;  % Default value
        fprintf('Using default significance level: %f\n', siglev);
    end

    % Add the vasca function to the path
    addpath('../matlab');

    % Call the vasca function
    vascao = vasca(parglmoVS, siglev);
    
    % Save results to JSON
    fid = fopen('vasca_matlab.json', 'w');
    if fid == -1
        error('Unable to open file to save results.');
    end
    fwrite(fid, jsonencode(vascao)); 
    fclose(fid)
end

function tf = isoctave()
    % Check if the environment is Octave
    tf = exist('OCTAVE_VERSION', 'builtin') ~= 0;
end

vasca_run();

File: tests/vasca_runners/vasca_run.R
# This script runs the vasca function and saves the results to a JSON file.
#
# It takes command-line arguments to specify the input data and optional parameters:
#
# Usage: Rscript vasca_run.R <parglmVS> <siglev>
#   - <parglmVS>: The output structure from the parglmVS function (in JSON format).
#   - <siglev>: Significance level (0.01 by default).
#
# Example usage:
#   Rscript vasca_run.R parglmVS_output.json 0.05
#   Rscript vasca_run.R parglmVS_output.json  # uses default siglev of 0.01
#
# Outputs:
#   - Saves the results as 'vasca_r.json'

# Install jsonlite if not installed
# Install jsonlite if not installed
if (!requireNamespace("jsonlite", quietly = TRUE)) {
    install.packages("jsonlite", repos = "https://cloud.r-project.org")
}

library(jsonlite)

# Main function
vasca_run <- function() {
    args <- commandArgs(trailingOnly = TRUE)
    
    if (length(args) < 1) {
        stop("Not enough arguments. Usage: Rscript vasca_run.R <parglmVS> <siglev>")
    }
    
    # Read input JSON file
    parglmVS_path <- args[1]
    if (!file.exists(parglmVS_path)) {
        stop("parglmVS must be a valid file.")
    }
    
    parglmVS <- fromJSON(parglmVS_path)
    cat("Loaded parglmVS from:", parglmVS_path, "\n")
    
    # Read siglev or use default value
    siglev <- ifelse(length(args) > 1, as.numeric(args[2]), 0.01)
    if (is.na(siglev)) {
        stop("The significance level must be a valid number.")
    }
    cat("Using significance level:", siglev, "\n")

    # Import vasca dependencies
    source('../R/vasca.R')
    source('../R/pcaEig.R')
    
    # Call the vasca function
    vascao <- vasca(parglmVS, siglev)
    
    # Save results to a JSON file
    write_json(vascao, "vasca_r.json", pretty = TRUE)
    cat("Results saved to vasca_r.json\n")
}

# Execute main function
vasca_run()
